### PR TITLE
[webdriver] Update tests

### PR DESCRIPTION
**Pull request by jugglinmike**
Recent changes to the "wptrunner" and "wd-client" projects have
invalidated some aspects of the WebDriver tests. Update the tests to
account for these changes, to correct previously-existing test bugs, and
to make the expectations more clear by using more explicit assertions.

@andreastt these tests currently hang in Firefox. Because the same WebDriver session
is re-used between tests, there are implicit navigation events between tests.
Marionette does not currently appears to be able to navigate between (for
instance) `about:blank#a` and `about:blank#b`. [I've filed a bug for that
here](https://bugzilla.mozilla.org/show_bug.cgi?id=1337464).

### CODE REVIEW

**Comment by jugglinmike** [comment]
Whoops, meant to mention @andreastt.

**Comment by w3c-bots** [comment]
# Chrome (unstable channel) #
Testing web-platform-tests at revision 5393e1aa6420564fc6a776a598eb00d9f9558aa8
Using browser at version 58.0.3000.4 dev
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
# Firefox (nightly channel) #
Testing web-platform-tests at revision 5393e1aa6420564fc6a776a598eb00d9f9558aa8
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
No tests run.

File: webdriver/navigation.py
Comment (jugglinmike) [inline webdriver/navigation.py]: This is a helpful fixture! This patch re-locates it to a separate module so we can re-use it from other tests. Those aren't included here, but I have some complete already that I plan to submit as a follow-up.

File: webdriver/util/wd_assert.py
Comment (andreastt) [inline webdriver/util/wd_assert.py]: I would use `types.StringTypes` since they could be unmarshalled to Unicode strings.

File: webdriver/util/wd_assert.py
Comment (andreastt) [inline webdriver/util/wd_assert.py]: `name` is misleading. Can you rename it `code`, `error_code` or similar?

Comment (andreastt) [inline webdriver/util/wd_assert.py]: Please add a docstring.

Comment (andreastt) [inline webdriver/util/wd_assert.py]: I think it would be useful if this function returned `result` unmodified.

File: webdriver/util/wd_assert.py
Comment (andreastt) [inline webdriver/util/wd_assert.py]: I would do a separate assertion that `value` exists on body.

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: Replace `[]` with `()`.

Also, why do you not call `session.execute` directly?

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: You can add `key="value"` to `send_command`.

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: I think this deserves some documentation since it modifies the DOM. To be honest, I’m not sure about the benefit of this as opposed to inlining a frame in a data URI in the test. That might be less surprising that having to check the fixtures to see what they actually do.

File: webdriver/util/wd_assert.py
Comment (andreastt) [inline webdriver/util/wd_assert.py]: I think this and the previous function, `error` should be renamed `assert_error` and `assert_success`. This file should also be renamed `asserts` in my opinion. We might want other custom assertions than purely WebDriver related ones.

File: webdriver/util/inline.py
Comment (andreastt) [inline webdriver/util/inline.py]: I have a more advanced version of this if you’re interested:

```py
def inline(doc, doctype="html", mime="text/html;charset=utf-8"):
    if doctype == "html":
        mime = "text/html;charset=utf-8"
    elif doctype == "xhtml":
        mime = "application/xhtml+xml"
        doc = r"""<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head>
    <title>XHTML might be the future</title>
  </head>

  <body>
    {}
  </body>
</html>""".format(doc)))
    return "data:{},{}".format(mime, urllib.quote(doc))
```

File: webdriver/conftest.py
Comment (andreastt) [inline webdriver/conftest.py]: What do these do?

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: Presumably this doesn’t need to be a fixture at all? You could make it a regular function and pass `session` to it?

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: What does this improve?

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: _Get Current URL_ should return an error. The old test here is wrong as it navigates. It should just be `session.url`.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: Doesn’t check for Unicode strings or that the body is JSON parsable?

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: You’ve already tested the raw command, so why not use `session.url` here?

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: Same here.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: This test has gone missing. The specification has changed since I wrote this test and it should instead return an ‘invalid argument’ error on passing a malformed URL.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: Why `dict`?

**Review by andreastt — CHANGES_REQUESTED** [review]

File: webdriver/navigation.py
Comment (jugglinmike) [inline webdriver/navigation.py]: This change makes the test more direct. The current factoring "reads" more like
a test for the `wdclient` library than for the specification itself--something
like, "Setting the `url` property on the `wdclient.Client` instance should
raise a Python exception." It also places more non-critical logic in the tested
code path.

My hope is that expressing the test subject in terms of `send_raw_command` will
make the test more legible to people who are familiar with the specification
but not the W3C's library--"Sending this command should return an error."

**Review by jugglinmike — COMMENTED** [review]
Thanks for your feedback, @andreastt! I've implemented most of your suggestions, though I do have some questions about a few things...

**Comment by jugglinmike** [comment]
I neglected to push up some of the changes you requested. Sorry about that;
they're all available now. I still have some questions about some of the finer
points of your earlier review, though. Would you mind taking another look?

File: webdriver/util/wd_assert.py
Comment (andreastt) [inline webdriver/util/wd_assert.py]: Well actually I guess the use cases for it would be very limited. Ignore by above request.

File: webdriver/util/wd_assert.py
Comment (andreastt) [inline webdriver/util/wd_assert.py]: You’re right, it does!

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: I buy the argument that invoking helpers in one, streamlined way is beneficial. This change is good.

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: Right, yes we do.

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: > could you explain why `()` is preferable here?

Arguments in Python are represented as tuples, so it is in harmony with the standard library to represent script arguments as tuples too.

File: webdriver/util/fixtures.py
Comment (andreastt) [inline webdriver/util/fixtures.py]: I think this is less surprising because there are fewer moving parts:

```py
document_el = session.execute("return document.documentElement")
frame_id = append_frame(document_el)

def append_frame(session):
    def x(parent_el = None):
        if parent_el is None:
            parent_el = session.execute("return document.documentElement")
        id = uuid.uuid4()

        return session.execute("""const [parentEl, id] = arguments;
            let frame = document.createElement("iframe");
            frame.setAttribute("id", id);
            parentEl.appendChild(frame);
            return frame;""", script_args=(parent_el, id))

    return x

wrapper = session.find.css("#wrapper")
new_frame = append_frame(wrapper)
print new_frame.get_attribute("id")
# => '17d3a557-c85a-4650-bfe9-0a8f3b624903'
```

But I’m still not convinced it is better than:

```py
def test_foo(session):
    session.navigate(inline("<iframe></iframe>"))
    iframe = session.find.css("iframe")
    …
```

File: webdriver/conftest.py
Comment (andreastt) [inline webdriver/conftest.py]: Makes sense.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: `basestring` appears to be fine, and `Result` does indeed parse the body.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: I don’t agree with you that `wd_assert.success(session.send_raw_command("GET", "url"), "about:blank")` is conceptually any more direct or simpler than making an HTTP request and asserting its characteristics, as we do in webdriver/contexts.py for Get Window Rect:

```py
with http.get("/url") as resp:
    assert resp.status == 200
    body = json.load(resp)

assert "value" in body
assert isinstance(body["value"], basestring)
assert body["value"] == "about:blank"
```

Whether you use `wd_assert.success` or the `session.url` property getter, they are both abstractions, and either one could repeat the above assertions. I subscribe to your ideal that we should avoid higher-level abstractions and promote immediacy and simplicity when testing a command, but I don’t see `wd_assert.success` as simple.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: navigation.py is intended to test the _Navigation_ chapter in the spec. When this file becomes long enough that it makes sense to split it out, we can do that.

File: webdriver/navigation.py
Comment (andreastt) [inline webdriver/navigation.py]: So I guess this is meant to switch to the frame? You could probably do {"element-6066-11e4-a52e-4f735466cecf": create_frame()}`.

**Comment by andreastt** [comment]
Sorry for my late response to this PR, but I’ve been sick and away from work for a week. Let me know if you need me to look at something in particular to unblock your work.

File: webdriver/util/fixtures.py
Comment (jugglinmike) [inline webdriver/util/fixtures.py]: That `append_frame` fixture has additional functionality compared to the
`create_frame` fixture submitted here. It generates a UUID and allows for the
specification of a parent element. This isn't necessarily a bad thing (although
neither is currently required by any tests), but it conflicts with your
statement about having "fewer moving parts." This makes me think I am
misunderstanding your point.

As for the use of `inline("<iframe></iframe")`, that approach depends on URL
modification and therefore cannot be used from tests that concern the value of
the current URL.

**Comment by andreastt** [comment]
Also sprach jugglinmike <notifications@github.com>:

> That append_frame fixture has additional functionality compared to
> the create_frame fixture submitted here. It generates a UUID and
> allows for the specification of a parent element. This isn't
> necessarily a bad thing (although neither is currently required by
> any tests), but it conflicts with your statement about having
> "fewer moving parts."   

Sure.  This is probably unneeded complexity.

> This makes me think I am misunderstanding your point.  As for the
> use of inline("<iframe></iframe"), that approach depends on URL
> modification and therefore cannot be used from tests that concern
> the value of the current URL.

That is a valid point.

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4756)

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4756)

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/4756)

**Comment by jugglinmike** [comment]
@ato I've rebased this branch and resolved the conflicts introduced via gh-5023 and gh-5024. Beyond that:

- The original version of this patch introduced a subdirectory named `util` for non-test files. Since the above patches introduces a subdirectory named `support` with a similar function, I relocated those files there.
- Incorporating feedback from your latest review
  - Miscellaneous improvements to the assertion functions as per:
    - https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100568459
    - https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100568595
    - https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100568990
    - https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100570202
  - Documented the `create_frame` Pytest fixture function as per https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100569435
  - Extended the `inline` helper function as per https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100571104
  - Updated the `create_frame` Pytest fixture function to use `Session#execute_script` as per https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100569066
  - Fixed an old test as per https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100572301
  - Re-introduced a mistakenly-deleted test as per https://github.com/w3c/web-platform-tests/pull/4756#discussion_r100573006

**Comment by jugglinmike** [comment]
@andreastt Commenting here again in case you missed my previous message.

**Comment by jugglinmike** [comment]
@andreastt Any word on this?

**Comment by jugglinmike** [comment]
@AutomatedTester Would you mind taking a look? I have a few patches lined up based on this work (they use some of the same code and they're formatted with similar patterns), so I'd like confirmation from a maintainer about this before moving forward.

**Review by andreastt — APPROVED** [review]
r+

**Pull request merged**
